node/coroutines/src/client_task.rs

```diff
@@ -269,28 +291,76 @@ impl ClientTask {
     }
 
     /// Tries importing block. If succeeds returns index of the block with the highest index.
+    /// If some blocks are missing, requests network to fetch them.
     fn try_import_block(
         &mut self,
+        peer_id: PeerId,
         beacon_block: SignedBeaconBlock,
         shard_block: SignedShardBlock,
     ) -> Option<BlockIndex> {
-        if let BlockImportingResult::Success { new_index } =
-            self.client.try_import_blocks(beacon_block, shard_block) {
+        self.peer_last_index.insert(peer_id, beacon_block.index());
+        // TODO: clonning here sucks, is there a better way?
+        match self.client.try_import_blocks(beacon_block.clone(), shard_block.clone()) {
+            BlockImportingResult::Success { new_index } => {
+                info!(target: "client",
+                    "Successfully imported block(s) up to {}, account_id={:?}",
+                    new_index, self.client.account_id
+                );
+                self.announce_block(beacon_block, shard_block);
+                Some(new_index)
+            },
+            BlockImportingResult::MissingParent { missing_indices, .. } => {
+                warn!(target: "client", "Missing indicies {:?}, account_id={:?}", missing_indices, self.client.account_id);
+                let (from_index, til_index) = (missing_indices.iter().min().unwrap(), missing_indices.iter().max().unwrap());
+                tokio::spawn(
+                    self.out_peer_tx.clone().send((peer_id, *from_index, *til_index)).map(|_| ()).map_err(|e| error!(target: "client", "Error sending request to fetch blocks from peer: {}", e))
+                );
+                None
+            },
+            // TODO: if block is invalid, we can extend `out_peer_tx` channel to also handle banning actions.
+            _ => None
+        }
+    }
 
-            info!(
-                "Successfully imported block(s) up to {}, account_id={:?}",
-                new_index, self.client.account_id
+    fn process_peer_state(&mut self, peer_id: PeerId, chain_state: ChainState) {
+        self.peer_last_index.insert(peer_id, chain_state.last_index);
+        if chain_state.last_index > self.client.beacon_chain.chain.best_index() {
+            // TODO: we should keep track of already fetching stuff.
```

Comment:
Addressing this comment. I think instead of keeping track of already fetching stuff we need to have some container called `PeersChainsInfo` which stores information about the chains from other peers. This container will then be updated from information coming from several sources, e.g.: (a) after-connection chain status, (b) periodic chain status, (c) shard block importing, (d) beacon block importing, etc. We then would have one task that monitors this container on an interval and our chain info and decides whether to send fetch requests. Otherwise, each channel that causes events (a),(b),(c),(d) need somehow collaborate together not to spam the network. Later we can modify the task that monitors `PeersChainsInfo` to react immediately on the changes in the container, but for now it can do a periodic check.

It also makes the subsequent implementation of proactive catchup easier. The very same task that monitors `PeersChainsInfo` can decide to periodically inquire the most recent "chain status" from some peers if they were not updated long enough.

Also `PeersChainsInfo`  and the corresponding task `PeersChainsInfoTask` seem like the logic that we can easily extract from `Client` and `ClientTask` without worrying too much about race conditions, because actions performed by `PeersChainsInfoTask` is mechanics that is completely separate from `Client->Nightshade` interactions. Also, our `Client` and `ClientTask` are growing quite a lot, and it seems like a safe logic to extract.

`PeersChainsInfo` can live inside `Client` as `Arc<RwLock< PeersChainsInfoTask >>` and `PeersChainsInfoTask` would monitor it together with the current state of the chains of the client.

Reply:
Created issue https://github.com/nearprotocol/nearcore/issues/708